ML guide initial draft (#38)
kdesai2018/ultimate-hackathon-starting-guide: machine-learning/README.md

# Machine Learning (Keras, PyTorch, Tensorflow)

## What is Machine Learning and Why Would I Use it?

> Machine Learning is a broad umbrella for methods of developing artificial intelligence, which is generally used to find patterns in large data sets based on mathematical and statistical calculations. This becomes very useful when a phenomena cannot be modeled by a deterministic algorithm.   

Reviewer: There're a lot of big words here. Can you reword it so that someone with no ML experience will understand it? Like when I read this I'm just like ehhh that kinda made sense but like I don't wanna sit here understanding it.

Keras (Recommended for beginners and small datasets):
  - High level API (Abstracts things more)
  - Slower runtimes(Shouldn't matter too much in this context)
  - Simplest to use
  - Least lines of code
  - Less accurate models
  - Simple models with minial debugging  
  
Tensorflow (Recommended for experienced users and high performance models despite smaller datasets):
  - High and Low level API's
  - Faster runtimes
  - Harder to use
  - Reduced code needed
  - Accurate models
  - Hard to debug  
  
 PyTorch (Recommended for machine learning gods and high performance models with large datasets):
  - Low Level API
  - Faster runtimes
  - Hardest to use
  - Most lines of code needed
  - Extremely accurate models
  - Good debugging support
  

## Pros and Cons
While ML is indeed very flashy, it also has some not-so-flashy aspects, which I will cover so you can consider if incorporating ML into your project is a good fit  
Pros:
  - Very relevant in the field of software development
  - Very flashy
  - Very powerful libraries
  - Generally beginner friendly

>   - Applicable to many projects  

Reviewer: Maybe talk a little more about what kind of projects it's applicable for. Imagine being a kid with an idea reading this and thinking hmm should I use this ML thing? And answer whatever questions would come to mind when thinking that

  - Needs large dataset (Machine Learning generally needs extremely large datasets to be accurate. In most cases, I have found myself without enough resources or enough time to create a custom dataset. It is also usually disappointing when your model doesn't correctly predict the category in front of the judges)
  - Reading thick documentation (Because of how powerful these libraries are, you will need to take some time to figure out which tools are the ones you need and which ones you don't)

## Takeaways
Machine Learning is a really useful tool for many hackathon projects if used effectively. 

## Resources I promised to link
1. [Official Tensorflow and Keras tutorial](https://www.tensorflow.org/tutorials)
2. [Keras Walkthrough](https://medium.com/@cyrilivargarcia/tensorflow-and-keras-a-beginners-tutorial-by-a-beginner-abd4c90f814f)
3. [Public Datasets](https://medium.com/towards-artificial-intelligence/the-50-best-public-datasets-for-machine-learning-d80e9f030279)

These sites are just a starting point - there is no substitute for getting your hands messy quickly. Fire up a venv or conda env (see note below) and get started! 

> ### What is venv and/or conda env

Reviewer: ```suggestion
### Where should I code this?
```

> Good question. We'll definitely try to include a more thorough explanation somewhere else in this repo, but heres the gist of it. 

Reviewer: ```suggestion
```

# Machine Learning (Keras, PyTorch, Tensorflow)

## What is Machine Learning and Why Would I Use it?

> Machine Learning is a broad umbrella for methods of developing artifial intelligence, which is generally used to find patterns in large data sets based on mathematical and statistical calculations. This becomes bery useful when a phenomena can not be modelled by a deterministic algorithm.   

Reviewer 2: ```suggestion
Machine Learning is a broad umbrella for methods of developing artificial intelligence, which is generally used to find patterns in large data sets based on mathematical and statistical calculations. This becomes very useful when a phenomena cannot be modeled by a deterministic algorithm.   

Before proceeding with this guide, check out [this](https://docs.google.com/presentation/d/1EImakverhNq-WhJwd-WIx2H04ZnCyUxsFk-9UcKoubQ/edit?usp=sharing) presentation made by Freetail Hackers. It's a quick overview about what AI is and a couple useful subfields within this space.
```

> In the context of a Hackathon, Machine Learning is designed to target a single problem with specific input parameters than can be consistently replicated on a large scale and thus, is most commonly used to develop an market solution tailored to a specific company challenge or a specific social problem. 

Reviewer 2: ```suggestion
In the context of a Hackathon, Machine Learning is designed to target a single problem with specific input parameters that can then be consistently replicated on a large scale and thus is most commonly used to develop an market solution tailored to a **specific company challenge or a specific social problem.** 
```

> The great thing about Machine Learning, is that it is applicable to most modern day problems. This ranges from economics, where ML is being used to developing models to predict stock prices (quantative analysis, a very lucrative field), to the medical field, where ML is being used to analyze X-Rays to determine if a patient has cancer.

Reviewer 2: ```suggestion
The great thing about Machine Learning is that it is applicable to most modern day problems. This ranges from economics, where ML is being used to developing models to predict stock prices (quantitative analysis, a very lucrative field), to the medical field, where ML is being used to analyze X-Rays to determine if a patient has cancer.
```

> Anything where normally we use our intution to predict a pattern, ML can be applied to do the work quicker and better!

Reviewer 2: ```suggestion
Anything where normally we use our intution to predict a pattern, ML can be applied to do the work quicker and better!
```
```suggestion
ML can be applied to solve a problem faster and more efficiently anywhere we would normally use our intuition to predict a pattern!
```

## Example Projects

> Here are a couple of the cool ML hackathon projects as some inspiration to get started!

Reviewer 2: ```suggestion
Here are some cool ML hackathon projects as inspiration to get started!
```

1. [NavAssistAI](https://devpost.com/software/navassistai) - This project won the Tensorflow World 2.0 Challenge, and attempts to help visually impaired individuals cross the street using machine learning. I like this project because its very applicable and usable in real life. 

2. [Coughvid](https://devpost.com/software/coughvid-ai-based-app-for-covid-19-screening-using-coughing) - This project placed 3rd at LauzHack for COVID. It uses ML to analyze coughing sounds and symptoms to predict if a patient has COVID-19.

> 3. [ColorMe](https://devpost.com/software/colorme-50fz68) - A unique project which usual Image Analysis to predict what color palettes for clothing would look best on the user.

Reviewer 2: ```suggestion
3. [ColorMe](https://devpost.com/software/colorme-50fz68) - A unique project which uses Image Analysis to predict what color palettes for clothing would look best on the user.
```

> 4. [StickyAR](https://devpost.com/software/stickyar) - Fun game built at HackMIT that takes a more light-hearted approach to ML. These hackers used OpenCV to detect sticky notes in a frame, and let the end user play a game using those sticky notes!

Reviewer 2: @jason-math sticky ar doesn't use AI at all, their entire platform is based on pretty simple image/video manipulation. maybe add a different example here?

There are a lot more! Browse [devpost](https://devpost.com) to see more cool projects. 

## Tools and libraries to get you started

> If you're gonna do any kind of ML project, I'd highly recommend Python. Many common Python ML libraries such as Keras, PyTorch, and Tensorflow abstract alot of the process for you, so you can focus on the bigger picture! This is especially good for beginners who may not understand everything that is going on behind the scenes just yet. C++ and Java are also pretty popular options, but I can't speak to their usability.

Reviewer 2: ```suggestion
If you're gonna do any kind of ML project, I'd highly recommend Python. Many common Python ML libraries such as Keras, PyTorch, and Tensorflow abstract a lot of the process (math) for you so that you can focus on the bigger picture! This is especially good for beginners who may not understand everything that is going on behind the scenes just yet. C++ and Java are also pretty popular options, but I can't speak to their usability.
```

Keras, PyTorch, and Tensorflow have a whole bunch of documentation out there for how to set them up, and how to use them. I'll link a couple python tutorials I found really useful down below. The best part about this library is that it works right out of the box (for the most part), allowing you to quickly move right on integrating it with the rest of the project.

You will most likely have to look into these frameworks to decide which one is the right fit for your project, but here are some general differences as a starting point:

Keras (Recommended for beginners and small datasets):
  - High level API (Abstracts things more)
  - Slower runtimes(Shouldn't matter too much in this context)
  - Simplest to use
  - Least lines of code
  - Less accurate models
  - Simple models with minial debugging  
  

> Tensorflow (Recommended for experienced users and high performance reduced size models):

Reviewer 2: maybe explain what you mean by "high performance reduced size models", even i'm not sure what that means

  - High and Low level API's
  - Faster runtimes
  - Harder to use
  - Reduced code needed
  - Accurate models
  - Hard to debug  
  
 PyTorch (Recommended for machine learning gods and high performance models with large datasets):
  - Low Level API
  - Faster runtimes
  - Hardest to use
  - Most lines of code needed
  - Extremely accurate models
  - Good debugging support
  

## Pros and Cons
While ML is indeed very flashy, it also has some not-so-flashy aspects, which I will cover so you can consider if incorporating ML into your project is a good fit  
Pros:
  - Very relevant in the field of software development
  - Very flashy
  - Very powerful libraries
  - Generally beginner friendly
  - Applicable to many projects  
  
Cons:
  - Long installation (You'll probably spend more time installing dependancies and libraries than actually working on your project)

>   - Needs large dataset (Machine Learning generally needs extremely large datasets to be accurate. In most cases, I have found myself       without enough resources or enough time to create a custom dataset. And it is usually disappointing when it doesn't correclt           predict in front of the judges)

Reviewer 2: ```suggestion
  - Needs large dataset (Machine Learning generally needs extremely large datasets to be accurate. In most cases, I have found myself without enough resources or enough time to create a custom dataset. It is also usually disappointing when your model doesn't correctly predict the category in front of the judges)
```

>   - Reading thick documentation (Because of how powerful these libraries are, you will need to take some time to figure out which tools     are the ones you need and which ones you don't)

Reviewer 2: ```suggestion
  - Reading thick documentation (Because of how powerful these libraries are, you will need to take some time to figure out which tools are the ones you need and which ones you don't)
```
